hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +48,272 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+        if (central_tag >= MIN_FLOPPY)
+            {
+            continue;
+            }
+        // if the central particle is not local, cannot read or write to it.
+        unsigned int central_idx = h_rtag.data[central_tag];
+        if (central_idx == NOT_LOCAL)
+            {
+            continue;
+            }
+        // collision on central particle itself already taken care of by collision rule
+        if (particle_index == central_idx)
+            {
+            continue;
+            }
+        // get velocities and masses
+        const Scalar4 vel_mass_const = h_velocity.data[particle_index];
+        const Scalar mass_const = vel_mass_const.w;
+        const quat<Scalar> orientation(h_orientation.data[central_idx]);
+        // get displacement
+        const Scalar4 postype_central = h_postype.data[central_idx];
+        const vec3<Scalar> pos_central(postype_central);
+        const Scalar4 postype_const = h_postype.data[particle_index];
+        const vec3<Scalar> pos_const(postype_const);
+        vec3<Scalar> displacement = pos_const - pos_central;
+
+        // update central particle velocity
+        // note: need to be sure that vector -> quaternion angular momentum conversion is correct
+        vec3<Scalar> linmom_change = mass_const
+                                     * (vec3<Scalar>(h_velocity.data[particle_index])
+                                        - vec3<Scalar>(h_initial_vel.data[idx]));
+        vec3<Scalar> angmom_change_vec = cross(displacement, linmom_change);
+        quat<Scalar> angmom_change = Scalar(2.0) * orientation * quat(0.0, angmom_change_vec);
+
+        // get the current accumulated momentum
+        vec3<Scalar> linmom_accum(h_linmom_accum.data[central_idx]);
+        quat<Scalar> angmom_accum(h_angmom_accum.data[central_idx]);
+        linmom_accum += linmom_change;
+        angmom_accum += angmom_change;
+        // update the alternative arrays
+        h_linmom_accum.data[central_idx]
+            = make_scalar3(linmom_accum.x, linmom_accum.y, linmom_accum.z);
+        h_angmom_accum.data[central_idx] = quat_to_scalar4(angmom_accum);
+        }
+    }
+
+void mpcd::CollisionMethod::finishRigidBodyCollision(uint64_t timestep)
+    {
+    ArrayHandle<Scalar3> h_linmom_accum(m_linmom_accum, access_location::host, access_mode::read);
+    ArrayHandle<Scalar4> h_angmom_accum(m_angmom_accum, access_location::host, access_mode::read);
+    ArrayHandle<Scalar4> h_velocity(m_pdata->getVelocities(),
+                                    access_location::host,
+                                    access_mode::readwrite);
+    ArrayHandle<Scalar4> h_angmom(m_pdata->getAngularMomentumArray(),
+                                  access_location::host,
+                                  access_mode::readwrite);
+
+    // add accumulated momentum to the central particle
+    unsigned int num_total = m_pdata->getN();
+    for (unsigned int idx = 0; idx < num_total; ++idx)
```

Comment:
You'll want to update this function to convert the space-frame angular momentum to the quaternion you need.

Just FYI: it isn't necessary to convert the `linmom_accum` to a `vec3`. You can use the native `Scalar3`, which has most of the math operations you're using already defined.

Last, I would add a check in this loop to only operate the central particles. That will save you a bunch of data loads and calculations if you do that up front.

Reply:
Actually, `cross` works only if both inputs are `vec3<Real>`

Reply:
`linmom_accum` and `angmom_accum` are both still `vec3` when they are calculated, but the conversion to `Scalar3` is now handled more elegantly.